Add benchmarking documentation (#50027)
dotnet/aspnetcore: docs/Benchmarks.md

- Run the updated Crank command line

## Benchmarking local changes

Use this technique when you want to benchmark a community contributed PR or when you don't have the changes available locally.

### Display all available benchmarks

- In the pull-request, add `/benchmark` in the comments.
- A text similar to this will be displayed (it might take up to 10 minutes):

---
Crank Pull Request Bot

`/benchmark <benchmark[,...]> <profile[,...]> <component,[...]> <arguments>`

Benchmarks:
- `plaintext`: TechEmpower Plaintext Scenario - ASP.NET Platform implementation
- `json`: TechEmpower JSON Scenario - ASP.NET Platform implementation
- `fortunes`: TechEmpower Fortunes Scenario - ASP.NET Platform implementation
- `yarp`: YARP - http-http with 10 bytes
- `mvcjsoninput2k`: Sends 2Kb Json Body to an MVC controller

Profiles:
- `aspnet-perf-lin`: INTEL/Linux 12 Cores
- `aspnet-perf-win`: INTEL/Windows 12 Cores
- `aspnet-citrine-lin`: INTEL/Linux 28 Cores
- `aspnet-citrine-win`: INTEL/Windows 28 Cores
- `aspnet-citrine-ampere`: Ampere/Linux 80 Cores
- `aspnet-citrine-amd`: AMD/Linux 48 Cores

> Components:

Reviewer: Clarify that components here basically refers to the "project under test" and that you won't get benchmarks correctly if your code modification isn't in the component.

Author: Added section explaining each part

- Run the updated Crank command line

## Benchmarking pull-requests

Use this technique when you want to benchmark a community contributed PR or when you don't have the changes available locally.

### Display all available benchmarks

- In the pull-request, add `/benchmark` in the comments.
- A text similar to this will be displayed (it might take up to 10 minutes):

---
Crank Pull Request Bot

`/benchmark <benchmark[,...]> <profile[,...]> <component,[...]> <arguments>`

Benchmarks:
- `plaintext`: TechEmpower Plaintext Scenario - ASP.NET Platform implementation
- `json`: TechEmpower JSON Scenario - ASP.NET Platform implementation
- `fortunes`: TechEmpower Fortunes Scenario - ASP.NET Platform implementation
- `yarp`: YARP - http-http with 10 bytes
- `mvcjsoninput2k`: Sends 2Kb Json Body to an MVC controller

Profiles:
- `aspnet-perf-lin`: INTEL/Linux 12 Cores
- `aspnet-perf-win`: INTEL/Windows 12 Cores
- `aspnet-citrine-lin`: INTEL/Linux 28 Cores
- `aspnet-citrine-win`: INTEL/Windows 28 Cores
- `aspnet-citrine-ampere`: Ampere/Linux 80 Cores
- `aspnet-citrine-amd`: AMD/Linux 48 Cores

Components:

> - `kestrel`
> - `mvc`

Reviewer 2: ```suggestion
- `kestrel` (Microsoft.AspNetCore.Server.Kestrel `src\Servers\Kestrel\build.cmd`)
- `mvc` (Microsoft.AspNetCore.Mvc `src\Mvc\build.cmd`)
- `routing` (Microsoft.AspNetCore.Routing `src\Http\build.cmd`)

See [prbenchmarks.aspnetcore.config.yml](https://github.com/aspnet/Benchmarks/blob/2bddb9b43ffbc8cfdb4a2c3777d4e5213bc42332/build/prbenchmarks.aspnetcore.config.yml) for more details.
```

Arguments: any additional arguments to pass through to crank, e.g. `--variable name=value`

---

Where:
- *Benchmark* is the application that will be stressed.
- *Profile* is the environment where the application will run on.

> - *Component* is the ASP.NET project under test that will be built and contains your changes.

Reviewer 2: I think we should be even more clear that changes outside of the "component" projects will not be benchmarked like @captainsafia suggested. I think that's the biggest footgun because it's not obvious when you get it wrong.

```suggestion
- *Component* is the ASP.NET project under test that will be built and contains your changes.

> [!IMPORTANT]  
> Any changes outside of the "component" projects will not be benchmarked.
```

# Benchmarks

> This document explains how to benchmark ASP.NET Core changes. It is intended for Microsoft employees.

Reviewer 3: ```suggestion
This document explains how to benchmark ASP.NET Core changes. Currently, only members of the .NET team at Microsoft have the permissions necessary to trigger performance runs on the official infrastructure.

```

Reviewer 4: > Currently, only members of the .NET team at Microsoft

We usually just say "the .NET team".

Aside -- at some future point I'd be curious to understand how close we are to being able to allow other contributors to run these jobs. ie., whether there's anything needs to be more robust to make it merely a resource/cost issue that we could tackle by maintaining a trusted group. (We can discuss offline sometime)

Author: This is a hard requirement since jobs are triggered on dnceng/internal. Allowing any external user to run anything in this ring would compromise secrets. So we only accept users that have the merge permission on this repository as this would have the same effect. So I would believe that this will be possible when we allow external contributors to merge PRs to main or release/x.

Reviewer 4: Gotcha

> This document explains how to benchmark ASP.NET Core changes. It is intended for Microsoft employees.

Reviewer 3: Should this say that the document explains how to benchmark changes _in the aspnetcore perf lab_?

> This document explains how to benchmark ASP.NET Core changes. It is intended for Microsoft employees.

Reviewer 4: I don't see anything about microbenchmarks 
1. the ones checked into dotnet/aspnetcore that are never run automatically
2. the ones in dotnet/performance -- do any of those run through code that exists in dotnet/aspnetcore

If crank supports any of these, and I hear it now supports dotnet/performance, we should link to instructions.
for the ones it doesn't support, we should at least point out they exist, and ideally link to instructions 

this is where the commentary on how to choose relevant benchmarks might come in handy. let's try to couch these instructions in terms that a new hire can use, who doesn't know about any of these things.

Reviewer 4: particularly since this doc is titled "how to benchmark changes" not "how to do system level benchmark of changes"

Author: The micro benchmarks in this repository were made mostly for investigations and not regression detections.

> It describes how to trigger benchmarks using the Crank CLI and the Pull-Request bot.

Reviewer 4: Consider a section here describing what PR's we imagine should use such runs? Essentially -- what kinds of changes tend to lead to perf regressions (that would be caught) and what would not benefit. This may also depend on how heavily the machines are subscribed -- that could feed into guidance here.

(If/when we set up a bot to post an encouragement/instructions into PR's that would most benefit from a perf run, perhaps by looking at files changed, that will obviously also help. Note that the labeler bot, eg., already looks at files changed in a PR.)

Author: I'd rather keep this document as succint as possible for devs who want to run a benchmark but don't remember how to. My personal opinion for which PRs should use such run is only "for external contributor PRs", and have .NET devs use locally built branches instead. I mentioned that in the dedicated paragraph.

> 

Reviewer 4: maybe here have a note that you're going to describe two alternatives -- local and PR -- below. they should choose which.

Reviewer 4: OK

## Benchmarking local changes

This is the recommended method as it is quick, flexible and doesn't require a pre-existing pull-request.

### Install Crank

> The Crank dotnet tool is required to start benchmarks remotely on the available performance infrastructure.

Reviewer 4: Maybe briefly explaining what Crank is -- what it can do, what it supports (eg., architectures), how long jobs typically take, ..

Personally I'm also curious where the name came from...

- At the bottom of the page there is table that lists the command lines that were used to execute this benchmark. Pick one.
- Run the command in a shell. This will start the benchmark remotely on the same target machines as the charts.

This command line contains the specific version of the runtimes and SDK that were used, which makes it deterministic. This means that if you want to investigate a regression you can pick two points on a chart and re-run these to confirm the regression. You can also change each version to understand if the regression is coming from ASP.NET or the .NET Runtime.

### Use local changes

Using the same Crank command line you can also test the impact of local changes.

- Build your changes in `release` for the desired architecture/os (if it matters).
- Add `--application.options.outputFiles c:\build\release` and replace the folder by the one containing your changes.
- Run the updated Crank command line

## Benchmarking pull-requests

> Use this technique when you want to benchmark a community contributed PR or when you don't have the changes available locally.

Reviewer 4: Is there anything I don't get from a PR run that I do get from a local run?

Author: If you are actively working on this PR you will need to push your changes and then wait more that 10 minutes for every change. Locally started benchmarks are much quicker to get results from, and don't need to push your code on GH. You also can't get profiles.

### Display all available benchmarks

- In the pull-request, add `/benchmark` in the comments.

> - A text similar to this will be displayed (it might take up to 10 minutes):

Reviewer 4: maybe link to a past PR that requested some runs this way, as an example?

- `aspnet-citrine-lin`: INTEL/Linux 28 Cores
- `aspnet-citrine-win`: INTEL/Windows 28 Cores
- `aspnet-citrine-ampere`: Ampere/Linux 80 Cores
- `aspnet-citrine-amd`: AMD/Linux 48 Cores

Components:
- `kestrel` (Microsoft.AspNetCore.Server.Kestrel `src\Servers\Kestrel\build.cmd`)
- `mvc` (Microsoft.AspNetCore.Mvc `src\Mvc\build.cmd`)
- `routing` (Microsoft.AspNetCore.Routing `src\Http\build.cmd`)

See [prbenchmarks.aspnetcore.config.yml](https://github.com/aspnet/Benchmarks/blob/2bddb9b43ffbc8cfdb4a2c3777d4e5213bc42332/build/prbenchmarks.aspnetcore.config.yml) for more details.

Arguments: any additional arguments to pass through to crank, e.g. `--variable name=value`

---

Where:
- *Benchmark* is the application that will be stressed.
- *Profile* is the environment where the application will run on.
- *Component* is the ASP.NET project under test that will be built and contains your changes.

> [!IMPORTANT]  
> Any changes outside of the "component" projects will not be benchmarked.

Create a new comment with the correct benchmarks, profiles, and components to build. For instance:

```console
/benchmark json aspnet-citrine-lin kestrel
```

> After up to 10 minutes a link should be displayed. Currently the results are not added to the PR anymore, you will need to follow the link and copy the results when the benchmark is finished.

Reviewer 3: Is there an issue for this that can be linked here?

Author: Currently adding more log in crank, you can approve the PR if you want to help. Right now there is no exception but nothing is in the comments either.

Reviewer 3: I helped! 😆

> After up to 10 minutes a link should be displayed. Currently the results are not added to the PR anymore, you will need to follow the link and copy the results when the benchmark is finished.

Reviewer 4: how do I know whether the results represent a regression? does it automatically do a baseline on HEAD at the same time? or i have to go compare/interpret the dashboard (that seems error prone given the complexity of the dashboard)

does it help me understand whether a diff might  be noise?

> After up to 10 minutes a link should be displayed. Currently the results are not added to the PR anymore, you will need to follow the link and copy the results when the benchmark is finished.

Reviewer 4: > After up to 10 minutes a link should be displayed

what should I do if it isn't, or I hit some other trouble? look at dotnet/crank docs? use some teams chat? (we're hopefully getting away from "ping Seb first thing" :))
